demos/ota/ota_demo_core_mqtt/ota_demo_core_mqtt.c

```diff
@@ -0,0 +1,1524 @@
+ *
+ * @param[in] pContext MQTT context which stores the connection.
+ * @param[in] pPublishInfo MQTT packet information which stores details of the
+ * job document.
+ */
+static void mqttJobCallback( MQTTContext_t * pContext,
+                             MQTTPublishInfo_t * pPublishInfo );
+
+/**
+ * @brief Callback that notifies the OTA library when a data block is received.
+ *
+ * @param[in] pContext MQTT context which stores the connection.
+ * @param[in] pPublishInfo MQTT packet that stores the information of the file block.
+ */
+static void mqttDataCallback( MQTTContext_t * pContext,
+                              MQTTPublishInfo_t * pPublishInfo );
+
+/**
+ * @brief callback to use with the MQTT context to notify incoming packet events.
+ *
+ * @param[in] pMqttContext MQTT context which stores the connection.
+ * @param[in] pPacketInfo Parameters of the incoming packet.
+ * @param[in] pDeserializedInfo Deserialized packet information to be dispatched by
+ * the subscription manager to event callbacks.
+ */
+static void mqttEventCallback( MQTTContext_t * pMqttContext,
+                               MQTTPacketInfo_t * pPacketInfo,
+                               MQTTDeserializedInfo_t * pDeserializedInfo );
+/*-----------------------------------------------------------*/
+
+void otaEventBufferFree( OtaEventData_t * const pxBuffer )
+{
+    if( sem_trywait( &bufferSemaphore ) == 0 )
+    {
+        pxBuffer->bufferUsed = false;
+        ( void ) sem_post( &bufferSemaphore );
+    }
+    else
+    {
+        LogError( ( "Failed to get buffer semaphore: "
+                    ",errno=%s",
+                    strerror( errno ) ) );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+OtaEventData_t * otaEventBufferGet( void )
+{
+    uint32_t ulIndex = 0;
+    OtaEventData_t * pFreeBuffer = NULL;
+
+    if( sem_trywait( &bufferSemaphore ) == 0 )
+    {
+        for( ulIndex = 0; ulIndex < otaconfigMAX_NUM_OTA_DATA_BUFFERS; ulIndex++ )
+        {
+            if( eventBuffer[ ulIndex ].bufferUsed == false )
+            {
+                eventBuffer[ ulIndex ].bufferUsed = true;
+                pFreeBuffer = &eventBuffer[ ulIndex ];
+                break;
+            }
+        }
+
+        ( void ) sem_post( &bufferSemaphore );
+    }
+    else
+    {
+        LogError( ( "Failed to get buffer semaphore: "
+                    ",errno=%s",
+                    strerror( errno ) ) );
+    }
+
+    return pFreeBuffer;
+}
+
+/*-----------------------------------------------------------*/
+
+static void otaAppCallback( OtaJobEvent_t event,
+                            void * pData )
+{
+    OtaErr_t err = OtaErrUninitialized;
+
+    /* OTA job is completed. so delete the MQTT and network connection. */
+    if( event == OtaJobEventActivate )
+    {
+        LogInfo( ( "Received OtaJobEventActivate callback from OTA Agent." ) );
+
+        if( pthread_mutex_lock( &mqttMutex ) != 0 )
```

Comment:
Doesn't this return 0 on success?
```suggestion
        if( pthread_mutex_lock( &mqttMutex ) == 0 )
```

Reply:
yes, thanks for pointing it out.